### DISCUSSION TITLE
Home DIR mapped via SSHFS disapper random.

### DISCUSSION THREAD

**Discussion by abixadamj** [ltsp/ltsp#316] [Q&A] [answered]
I have few installations based on Linux Mint 19.3, 
My problem is that in random time when users are logged in, the home dir is UNMAPPED in client, so user does not acces to /home/_user_ dir.
I don't have any ide, what can cause this, what could I check? I have about 20 installations around in Poland, but this problem occurs only in 3 last instalations :

```
root@GRYFINO-SP3:~# uname -a
Linux GRYFINO-SP3 4.15.0-117-generic ltsp/ltsp#118-Ubuntu SMP Fri Sep 4 20:02:41 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
root@GRYFINO-SP3:~# ltsp info
LTSP version: 20.06-1~ubuntu18.04.1
SERVER XSESSIONS: mate.desktop
SERVER OS: PRETTY_NAME="Linux Mint 19.3"
================
root@ceglow-free-serwer:~# uname -a
Linux ceglow-free-serwer 4.15.0-117-generic ltsp/ltsp#118-Ubuntu SMP Fri Sep 4 20:02:41 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
root@ceglow-free-serwer:~# ltsp info
LTSP version: 20.06-1~ubuntu18.04.1
SERVER XSESSIONS: mate.desktop
SERVER OS: PRETTY_NAME="Linux Mint 19.3"
================
root@bytom-free-serwer:~# uname -a
Linux bytom-free-serwer 4.15.0-112-generic ltsp/community#96-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
root@bytom-free-serwer:~# ltsp info
LTSP version: 20.06-1~ubuntu18.04.1
SERVER XSESSIONS: mate.desktop
SERVER OS: PRETTY_NAME="Linux Mint 19.3"
```
Different kernels and the same LTSP. @alkisg  - maybe you have heard about similar problems...
Adam

**Comment by alkisg** [comment]
Hi Adam, no, I haven't heard of any issues with sshfs connections getting dropped.
You can run `dmesg` on the client to see if its networking was disconnected for a long time.
Also `journalctl -u ssh` on the server might have some clues.

If networking itself isn't the problem, then check if tcp keepalive is. Try to run ssh from a client or even from another machine to the server, then leave it without activity for an hour, and see if it's disconnected.

**Comment by abixadamj** [comment]
Nothing interested. I remove all like

```
wrz 10 09:40:48 GRYFINO-SP3 sshd[21125]: Did not receive identification string from 192.168.3.1 port 46526
wrz 10 09:40:50 GRYFINO-SP3 sshd[21126]: Did not receive identification string from 192.168.3.1 port 46725
wrz 10 09:40:52 GRYFINO-SP3 sshd[21129]: Did not receive identification string from 192.168.3.1 port 46938
wrz 10 09:40:54 GRYFINO-SP3 sshd[21135]: Did not receive identification string from 192.168.3.1 port 47141
wrz 10 09:40:56 GRYFINO-SP3 sshd[21136]: Did not receive identification string from 192.168.3.1 port 47336
wrz 10 09:40:58 GRYFINO-SP3 sshd[21137]: Did not receive identification string from 192.168.3.1 port 47531
```

and got only:

```
wrz 10 11:22:29 GRYFINO-SP3 sshd[4528]: Failed password for u17k8a from 192.168.3.237 port 51804 ssh2
wrz 10 11:22:29 GRYFINO-SP3 sshd[4528]: Connection closed by authenticating user u17k8a 192.168.3.237 port 51804 [preauth]
wrz 10 11:22:49 GRYFINO-SP3 sshd[4560]: Accepted password for u17k8a from 192.168.3.237 port 51806 ssh2
wrz 10 11:22:49 GRYFINO-SP3 sshd[4560]: pam_unix(sshd:session): session opened for user u17k8a by (uid=0)
wrz 10 11:23:02 GRYFINO-SP3 sshd[4645]: Accepted password for u13k8a from 192.168.3.236 port 36124 ssh2
wrz 10 11:23:02 GRYFINO-SP3 sshd[4645]: pam_unix(sshd:session): session opened for user u13k8a by (uid=0)
wrz 10 11:23:02 GRYFINO-SP3 sshd[4716]: Accepted password for u10k8a from 192.168.3.223 port 36468 ssh2
wrz 10 11:23:02 GRYFINO-SP3 sshd[4716]: pam_unix(sshd:session): session opened for user u10k8a by (uid=0)
wrz 10 11:23:08 GRYFINO-SP3 sshd[4799]: Connection closed by authenticating user u20k8a 192.168.3.222 port 44216 [preauth]
wrz 10 11:23:25 GRYFINO-SP3 sshd[4819]: Accepted password for u02k8a from 192.168.3.231 port 55410 ssh2
wrz 10 11:23:25 GRYFINO-SP3 sshd[4819]: pam_unix(sshd:session): session opened for user u02k8a by (uid=0)
wrz 10 11:23:48 GRYFINO-SP3 sshd[4925]: Accepted password for u11k8a from 192.168.3.239 port 54260 ssh2
wrz 10 11:23:48 GRYFINO-SP3 sshd[4925]: pam_unix(sshd:session): session opened for user u11k8a by (uid=0)
wrz 10 11:24:22 GRYFINO-SP3 sshd[5028]: Accepted password for u07k8a from 192.168.3.235 port 60230 ssh2
wrz 10 11:24:22 GRYFINO-SP3 sshd[5028]: pam_unix(sshd:session): session opened for user u07k8a by (uid=0)
wrz 10 11:24:22 GRYFINO-SP3 sshd[5028]: pam_unix(sshd:session): session closed for user u07k8a
wrz 10 11:24:26 GRYFINO-SP3 sshd[5104]: Connection closed by 127.0.0.1 port 49652 [preauth]
wrz 10 11:25:15 GRYFINO-SP3 sshd[5147]: Accepted password for u03k8a from 192.168.3.234 port 57766 ssh2
wrz 10 11:25:15 GRYFINO-SP3 sshd[5147]: pam_unix(sshd:session): session opened for user u03k8a by (uid=0)
wrz 10 11:25:20 GRYFINO-SP3 sshd[5231]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=192.168.3.222  user=u20k8a
```
I have there:
```
root@GRYFINO-SP3:~# cat /etc/ssh/sshd_config | grep TCP
#TCPKeepAlive yes
 ```
I will remove `#` and will see....

```
root@GRYFINO-SP3:~# cat /etc/ssh/sshd_config | grep -v ^# | uniq

ChallengeResponseAuthentication no

UsePAM yes

X11Forwarding yes
PrintMotd no
TCPKeepAlive yes
ClientAliveInterval 10

AcceptEnv LANG LC_*

Subsystem	sftp	/usr/lib/openssh/sftp-server

root@GRYFINO-SP3:~# service ssh restart
```

**Comment by abixadamj** [comment]
I have found:
https://serverfault.com/questions/306796/sshfs-problem-when-losing-connection
There is:
If you are mounting the filesystem with `-o reconnect` (See the sshfs man page) the system should reconnect on its own.

@alkisg -where in ltsp/client dir shoult I check mount options?

**Comment by alkisg** [comment]
-o reconnect means that you're using passwordless ssh, which we're not; in LTSP, we're using passwords for authentication, so reconnect won't work.

Try to focus on why you get ssh disconnections while noone else does. Run `ssh server`, let it there for a couple of hours, and try to troubleshoot why it doesn't stay connected.

I'll move this issue to the community wiki as it's not a problem with the ltsp code.

**Comment by abixadamj** [comment]
OK, I will check in next few hours. I go to my client.

**Comment by alkisg** [comment]
In a setup where SSH had similar issues for me, I had to do use this:

```shell
ssh -o TCPKeepAlive=no -o ServerAliveInterval=60 ...
```

This setup was not using LTSP, and I think the problem was caused by a pfSense installation somewhere between the source and destination computer.

**Comment by abixadamj** [comment]
I did not find any interesting. I cannot reproduce an error. I remove latest kernel 4.15.0-117 and change to 4.15.0-112.
I will check in next days...

**Comment by abixadamj** [comment]
@alkisg and everyone who could be interested in issue.
I have reinstall all the server twice:
1. In client's school - Mint 19.2 with 4.15.0.78 kernel + 20 Clients - there were errors sometimes on random PCs
2. In my home - Mint 20 with 5.4 kernel + 1 client - after 20 minutes of working there was an error

So - the problem MUST be in hardware, I thought - and I found - Ethernet card in server. Found is just like - I think eth must be a culprit. I had disabled it (Dell on-board) and put simple PCIExpress TP-Link ethernet card.

IT WORKS !!!!!!

I wrote simple Python script running in loop:
* read content of txt file from sshfs home dir
* write that content to random named file to another dir on sshfs dir

In about 2 hours script did about 2 millions (2 000 000 000) reads and writes via sshf - **NO ERROR!!!!**

So, I go to client's school and install server there and make the test - running those script on 20 PCs simultaneously ...
Then I will write you about results.

**Comment by abixadamj** [comment]
I have been at school. Everything WORKS !!! I can close this ticket. The problem were at Ethernet card on server (hardware).

**Comment by abixadamj** [comment]
I want to reopen this issue, because I think, it is kernel/ethernet module related. I just install fresh Mint 20 with:

```
root@demo-serwer:~# uname -a
Linux demo-serwer 5.4.0-48-generic ltsp/ltsp#216-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
root@demo-serwer:~# lsb_release -a
LSB Version:	core-11.1.0ubuntu2-noarch:printing-11.1.0ubuntu2-noarch:security-11.1.0ubuntu2-noarch
Distributor ID:	Linuxmint
Description:	FREE-Desktop 2021 (0.01a)
Release:	20
Codename:	ulyana
```

And only 1 Dell Core i5 as a client with gigabit Tp-Link Archer router dhcp server. Only server and client connected to the router. Client boot up with 10 seconds, but 10 seconds after login $HOME (mapped via sshFS) disappeared... and after few seconds come back, after next few disappeared again... 

The laptop is OK, I work on it in every day....

```
root@demo-serwer:~# lshw -c net
  *-network                 
       description: Ethernet interface
       product: Ethernet Connection I218-LM
       vendor: Intel Corporation
       physical id: 19
       bus info: pci@0000:00:19.0
       logical name: eno1
       version: 04
       serial: f0:1f:af:70:1a:b6
       size: 1Gbit/s
       capacity: 1Gbit/s
       width: 32 bits
       clock: 33MHz
       capabilities: pm msi bus_master cap_list ethernet physical tp 10bt 10bt-fd 100bt 100bt-fd 1000bt-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=e1000e driverversion=3.2.6-k duplex=full firmware=0.7-3 ip=192.168.0.100 latency=0 link=yes multicast=yes port=twisted pair speed=1Gbit/s
       resources: irq:51 memory:f7e00000-f7e1ffff memory:f7e3c000-f7e3cfff ioport:f080(size=32)
  *-network DISABLED
       description: Wireless interface
       product: Wireless 7260
       vendor: Intel Corporation
       physical id: 0
       bus info: pci@0000:02:00.0
       logical name: wlp2s0
       version: 73
       serial: fc:f8:ae:8f:41:c0
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list ethernet physical wireless
       configuration: broadcast=yes driver=iwlwifi driverversion=5.4.0-48-generic firmware=17.3216344376.0 latency=0 link=no multicast=yes wireless=IEEE 802.11
       resources: irq:49 memory:f7d00000-f7d01fff

root@demo-serwer:~# dmesg | grep e1000
[    0.781043] e1000e: Intel(R) PRO/1000 Network Driver - 3.2.6-k
[    0.781044] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[    0.795070] e1000e 0000:00:19.0: Interrupt Throttling Rate (ints/sec) set to dynamic conservative mode
[    1.152791] e1000e 0000:00:19.0 0000:00:19.0 (uninitialized): registered PHC clock
[    1.221364] e1000e 0000:00:19.0 eth0: (PCI Express:2.5GT/s:Width x1) f0:1f:af:70:1a:b6
[    1.221366] e1000e 0000:00:19.0 eth0: Intel(R) PRO/1000 Network Connection
[    1.221407] e1000e 0000:00:19.0 eth0: MAC: 11, PHY: 12, PBA No: FFFFFF-0FF
[    1.222309] e1000e 0000:00:19.0 eno1: renamed from eth0
[   11.880281] e1000e: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
[  665.239089] e1000e: eno1 NIC Link is Down
[ 2278.987026] e1000e: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: Tx
[ 6799.198207] e1000e: EEE TX LPI TIMER: 00000011
[ 6801.276428] e1000e: eno1 NIC Link is Down
[ 6806.663466] e1000e: eno1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: Tx
```
Down and up are because it was sleeping...

I think that e1000 kernel drivers is responsible for my problems...

**Comment by abixadamj** [comment]
Very !!! Strange situation.
The place where I have changed card on server is still a problem. I turn off Intel onboard and put Realtek PCI-E.
But the problem still exist.

In random time, on random PC, $HOME_DIR mounted by SSHFS disappears ! There is tlii a mount line in `/proc/mount`, but when the user want to open icon on Desktop or document... there are no files.

I have installed server on my private laptop (Dell) and connect to Dell PC as a client. For 3 times I have testing:
2 times everything was OK (Python script wrote 10000 files on $HOME_DIR)
1 time after 3 seconds from login $HOME_DOR disappers !

Of course - there are NO ERRORS in syslog or auth.log.

@alkisg - did you have any situation like this? Maybe someone else?

**Comment by abixadamj** [comment]
Next comment - I will treat this place like a diary ;-) 
Friend, who is a teacher in that school, tested...

1. The problem is often after Mozilla Firefox is running.
2. Sometimes even reboot the client does not solve a problem, after reboot and login there is no connection to $HOME_DIR just in a few seconds after login!

@alkisg - I have a question - what do you think - do yo see a way to test mounting HOME in login process and sent this log to a server?
Maybe in file `/usr/share/ltsp/client/login/pamltsp`
I see there something about login and mount maybe somewhere in `pam_auth()` function?
What do you think?

**Comment by alkisg** [comment]
Adam, you didn't do my suggestions from https://github.com/ltsp/community/issues/158#issuecomment-690204549.

So again I would suggest that you boot an LTSP client without LTSP.
Boot it from a live CD.
Connect to the server via ssh.
See if it stays connected for 3 hours.
If not, it's a networking problem that is completely unrelated to LTSP.
Keep things simple. Troubleshoot this first.

**Comment by abixadamj** [comment]
OK, I will try to do it. It is hard, because I am not on a place (it is 60 km from me) and only 2 days in a week there are lessons in that classroom. But, I will do this test in my home, where I have 2 PC's and where this problem was few times.

**Comment by abixadamj** [comment]
OK, I have my 2 PCs and TpLink router between them.
![Zrzut ekranu z 2020-10-09 13-02-12](https://user-images.githubusercontent.com/5088643/95576050-127fb780-0a30-11eb-96eb-ea4f1eddda77.png)

I have mounted via sshfs directory from server to client. And I will let it for at least 2 hours.
![Zrzut ekranu z 2020-10-09 13-02-33](https://user-images.githubusercontent.com/5088643/95576133-3216e000-0a30-11eb-82ae-30433c9700fa.png)

Then I will write.

**Comment by abixadamj** [comment]
@alkisg 
2 hours and still works.

![Zrzut ekranu z 2020-10-09 15-14-57](https://user-images.githubusercontent.com/5088643/95587522-6c897880-0a42-11eb-93d7-15099e218e53.png)

Now, I have to shut down PCs and go out for 2 days, but on Monday I will start the same test for at least 5 hours.

**Comment by alkisg** [comment]
The networking problem is at school.
Proving that you do not have a network problem at home doesn't help.
Test at school, not at home.

Anyway, as no other LTSP users have that problem, I'm pretty sure that it's not related to LTSP, so I will not comment on this issue any more. Please ask a technician to help you solve the school networking problem. Cheers!

**Comment by abixadamj** [comment]
Dear @alkisg .

I'm grateful for your attention and help. Your words and thoughts push me to more tests.
This is what I found.
----
After 5 hours of testing sshfs (of course everything was ok), I discovered my source of problem.

![Zrzut ekranu z 2020-10-11 20-25-37](https://user-images.githubusercontent.com/5088643/95689251-b1aed580-0c0f-11eb-968c-53f18529dffb.png)


**Web browser (Firefox, or Google Chrome, or Opera) - when I start it, the HOME_DIR is inaccessible. 
But, when I close web browser - everything comes to normal work.**
It is 100% repetitive - at least in my server home. So I investigate browser witch `sctrace`, and I found:

```
socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0) = 32
connect(32, {sa_family=AF_UNIX, sun_path=@"/tmp/.X11-unix/X0"}, 20) = 0
getpeername(32, {sa_family=AF_UNIX, sun_path=@"/tmp/.X11-unix/X0"}, [124->20]) = 0
uname({sysname="Linux", nodename="ltsp100", ...}) = 0
access("/home/uczniowie/k9/u01k9/.Xauthority", R_OK) = 0
openat(AT_FDCWD, "/home/uczniowie/k9/u01k9/.Xauthority", O_RDONLY) = 33
fstat(33, {st_mode=S_IFREG|0600, st_size=52, ...}) = 0
read(33, "\1\0\0\7ltsp100\0\0010\0\22MIT-MAGIC-COOKIE"..., 4096) = 52
read(33, "", 4096)                      = 0
close(33)                               = 0
getsockname(32, {sa_family=AF_UNIX}, [124->2]) = 0
fcntl(32, F_GETFL)                      = 0x2 (flags O_RDWR)
fcntl(32, F_SETFL, O_RDWR|O_NONBLOCK)   = 0
fcntl(32, F_SETFD, FD_CLOEXEC)          = 0
poll([{fd=32, events=POLLIN|POLLOUT}], 1, -1) = 1 ([{fd=32, revents=POLLOUT}])
writev(32, [{iov_base="l\0\v\0\0\0\22\0\20\0\0\0", iov_len=12}, {iov_base="", iov_len=0}, {iov_base=
"MIT-MAGIC-COOKIE-1", iov_len=18}, {iov_base="\0\0", iov_len=2}, {iov_base="\220jW\3013\346a\224X\26
2\323\261\205|\307\372", iov_len=16}, {iov_base="", iov_len=0}], 6) = 48
recvfrom(32, 0x179147127520, 8, 0, NULL, NULL) = -1 EAGAIN (Zasoby chwilowo niedostępne)
```

As you can see, it reads the **MIT-MAGIC-COOKIE** and then ....**EAGAIN** (resources temporary inaccessible)

I don't want you to comment this - **I know, it is not LTSP related problem.**  But, maybe some other guys will search for similar problems, please, do not delete this issue. I will search more on this and when I found, what I do wrong with my server, I will write it here and then close this issue.

**Comment by alkisg** [comment]
> I know, it is not LTSP related problem ... please, do not delete this issue.

Hello! Do not worry about that, the LTSP community discussions are open for all topics that are somewhat related to LTSP, even if they're not caused by LTSP itself.
I only said that since it's not an LTSP issue, I do not currently have enough free time to personally help in troubleshooting it. Others are surely welcome to help you.

**Comment by abixadamj** [comment]
Alkis, you have helped me a lot ;-) I know very well that time is worth a gold in that time. I will investigate this issue by myself ;-)

**Comment by abixadamj** [comment, marked as answer]
@alkisg - I think I have found.
`var/cache/*` entry in file `/etc/ltsp/add-image.excludes`

One whole day of searching, installing, running ;-)

**Comment by abixadamj** [comment]
I can close this, as I found... If anyone will have such behavior - please read and try to correct.

**Comment by alkisg** [comment]
Do you mean that you added the following in ltsp.conf?

```shell
ADD_IMAGE_EXCLUDES="var/cache/*"
```

This would cause a lot of problems, as the directories in /var/cache then would all be omitted from the image, and some services wouldn't be allowed to write there.

**Reply by abixadamj** [reply]
Yes, that is why I had so many problems ;-) Please, don't ask, why I put it there. I don't remember. Maybe I wanted to make some lighter image ;-( Bad Idea ... But now everything is working great again! @alkisg - thank you so much for your patient and time.
